pkg/agent/controller/networkpolicy/l7engine/l7visbility_watcher.go

```diff
@@ -0,0 +1,393 @@
+// Copyright 2023 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package l7engine
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/agent/config"
+	"antrea.io/antrea/pkg/apis/crd/v1alpha2"
+
+	"antrea.io/antrea/pkg/agent/controller/trafficcontrol"
+)
+
+var (
+	L7mutex sync.Mutex
+)
+
+const (
+	watcherName = "L7Watcher"
+)
+
+type l7Watcher struct {
+	tcController  *trafficcontrol.Controller
+	clientset     kubernetes.Interface
+	l7Reconciler  *Reconciler
+	labelsToTCMap map[string][]string
+}
+
+func NewL7Watcher(tcController *trafficcontrol.Controller, clientset kubernetes.Interface) *l7Watcher {
+	l7Reconciler := NewReconciler()
+	labelsToTCMap := make(map[string][]string)
+	return &l7Watcher{
+		tcController:  tcController,
+		clientset:     clientset,
+		l7Reconciler:  l7Reconciler,
+		labelsToTCMap: labelsToTCMap,
+	}
+}
+
+func (l7w *l7Watcher) Run(stopCh <-chan struct{}) {
+	klog.InfoS("Starting", "Watcher", watcherName)
+	defer klog.InfoS("Shutting down", "Watcher", watcherName)
+	// Create a watcher for pods and namespaces
+	podWatcher, err := l7w.clientset.CoreV1().Pods("").Watch(context.TODO(), metav1.ListOptions{
+		FieldSelector: fields.Everything().String(),
+	})
+	if err != nil {
+		klog.ErrorS(err, "")
+	}
+	defer podWatcher.Stop()
+
+	namespaceWatcher, err := l7w.clientset.CoreV1().Namespaces().Watch(context.TODO(), metav1.ListOptions{
+		FieldSelector: fields.Everything().String(),
+	})
+	if err != nil {
+		klog.ErrorS(err, "")
+	}
+	defer namespaceWatcher.Stop()
+
+	// Set up channels to receive watch events
+	podEvents := podWatcher.ResultChan()
+	namespaceEvents := namespaceWatcher.ResultChan()
+
+	// Start listening for events
+	go l7w.processPodEvents(podEvents)
+	go l7w.processNamespaceEvents(namespaceEvents)
+
+	<-stopCh
+}
+
+func (l7w *l7Watcher) processPodEvents(events <-chan watch.Event) {
+	direction := v1alpha2.DirectionBoth
+	for event := range events {
+		pod, ok := event.Object.(*corev1.Pod)
+		if !ok {
+			klog.InfoS("Received unexpected object: ", "", event.Object)
+			continue
+		}
+		nodeName := pod.Spec.NodeName
+		podLabels := pod.ObjectMeta.Labels
```

Comment:
So this gets all the pod labels, but then are used as pod selectors? I'm a little confused on how they are equivalent at line 258.

Reply:
selectors are for the Traffic control CRD, the labels are passed to the createTrafficControl func to be used as selectors in CRD